File: sim.py
from argparse import ArgumentParser

# Smith predictor: A simple counter-based predictor.
# It uses a saturating counter to predict whether a branch will be taken ("t") or not taken ("n").
class Smith:
    def __init__(self, b):
        self.b = b  # Number of bits in the counter
        self.counter = 1 << (b - 1)  # Initialize counter to mid-point
        self.thres = 1 << (b - 1)  # Threshold for predicting "t"

    def predict(self, pc):
        # Predict "t" if counter >= threshold, otherwise predict "n"
        if self.counter >= self.thres:
            return "t"
        else:
            return "n"
        
    def update(self, outcome):
        # Update the counter based on the actual outcome
        if outcome == "t":
            self.counter = min(self.counter + 1, (1 << self.b) - 1)  # Increment counter
        else:
            self.counter = max(self.counter - 1, 0)  # Decrement counter

# Bimodal predictor: Uses a table of counters indexed by the program counter (PC).
# Each counter predicts "t" or "n" based on its value.
class Bimodal:
    def __init__(self, m):
        self.m = m  # Number of index bits
        self.table = [4] * (1 << m)  # Initialize table with mid-point values
        self.index = 0

    def predict(self, pc):
        # Calculate the index using bits m+1 to 2 of the PC
        self.index = (pc >> 2) & ((1 << self.m) - 1)

        # Predict "t" if counter >= 4, otherwise predict "n"
        if self.table[self.index] >= 4:
            return "t"
        else:
            return "n"

    def update(self, outcome):
        # Update the counter at the calculated index
        if outcome == "t":
            self.table[self.index] = min(self.table[self.index] + 1, 7)  # Increment counter
        else:
            self.table[self.index] = max(self.table[self.index] - 1, 0)  # Decrement counter

# Gshare predictor: Combines global history with the PC to index into a table of counters.
class Gshare:
    def __init__(self, m, n):
        self.m = m  # Number of index bits
        self.n = n  # Number of global history bits
        self.table = [4] * (1 << m)  # Initialize table with mid-point values
        self.global_history = 0  # Initialize global history
        self.idx = 0

    def predict(self, pc):
        # Calculate the index by XORing the PC and global history
        self.idx = (pc >> 2) & ((1 << self.m) - 1)
        self.idx = self.idx ^ self.global_history

        # Predict "t" if counter >= 4, otherwise predict "n"
        if self.table[self.idx] >= 4:
            return "t"
        else:
            return "n"

    def update(self, outcome):
        # Update the counter at the calculated index
        if outcome == "t":
            self.table[self.idx] = min(self.table[self.idx] + 1, 7)  # Increment counter
        else:
            self.table[self.idx] = max(self.table[self.idx] - 1, 0)  # Decrement counter
        
        # Update the global history
        self.global_history = (self.global_history >> 1) | (int(outcome == "t") << (self.n - 1))

# Hybrid predictor: Combines Gshare and Bimodal predictors using a chooser table.
class Hybrid:
    def __init__(self, k, m1, n, m2):
        self.k = k  # Number of chooser table index bits
        self.gshare = Gshare(m1, n)  # Gshare predictor
        self.bimodal = Bimodal(m2)  # Bimodal predictor
        self.chooser_table = [1] * (1 << k)  # Initialize chooser table
        self.idx = 0

        self.gshare_output = None
        self.bimodal_output = None

    def predict(self, pc):
        # Calculate the chooser table index
        self.idx = (pc >> 2) & ((1 << self.k) - 1)

        # Get predictions from Gshare and Bimodal
        self.gshare_output = self.gshare.predict(pc)
        self.bimodal_output = self.bimodal.predict(pc)

        # Use chooser table to select the predictor
        if self.chooser_table[self.idx] >= 2:
            return self.gshare_output
        else:
            return self.bimodal_output

    def update(self, outcome):
        # Update the predictor used
        if self.chooser_table[self.idx] >= 2:
            self.gshare.update(outcome)
        else:
            self.bimodal.update(outcome)
            # Also update the global history of the Gshare predictor
            self.gshare.global_history = (self.gshare.global_history >> 1) | (int(outcome == "t") << (self.gshare.n - 1))

        # Update the chooser table based on prediction accuracy
        if self.gshare_output == outcome and self.bimodal_output != outcome:
            self.chooser_table[self.idx] = min(self.chooser_table[self.idx] + 1, 3)
        elif self.gshare_output != outcome and self.bimodal_output == outcome:
            self.chooser_table[self.idx] = max(self.chooser_table[self.idx] - 1, 0)

if __name__ == "__main__":
    # Main code: Parses arguments, initializes the predictor, and processes the trace file.
    parser = ArgumentParser(description="Branch Prediction")
    parser.add_argument("predictor", type=str, help="Predictor type")
    parser.add_argument("arg1", type=str, help="Argument 1")
    parser.add_argument("arg2", type=str, nargs='?', default=None, help="Argument 2")
    parser.add_argument("arg3", type=str, nargs='?', default=None, help="Argument 3")
    parser.add_argument("arg4", type=str, nargs='?', default=None, help="Argument 4")
    parser.add_argument("arg5", type=str, nargs='?', default=None, help="Argument 5")

    args = parser.parse_args()
    trace_file = None

    # Initialize the appropriate predictor based on the input arguments
    if args.predictor == "smith":
        predictor = Smith(int(args.arg1))
        trace_file = args.arg2
    elif args.predictor == "gshare":
        predictor = Gshare(int(args.arg1), int(args.arg2))
        trace_file = args.arg3
    elif args.predictor == "bimodal":
        predictor = Bimodal(int(args.arg1))
        trace_file = args.arg2
    elif args.predictor == "hybrid":
        predictor = Hybrid(int(args.arg1), int(args.arg2), int(args.arg3), int(args.arg4))
        trace_file = args.arg5
    else:
        print("Invalid predictor type")
        exit(1)

    number_branches = 0
    number_mispredictions = 0
    print("COMMAND")
    print("./sim", end="")
    for arg in vars(args).values():
        if arg is not None:
            print(f" {arg}", end="")
    print("\nOUTPUT")
    # Process the trace file
    with open(trace_file, "r") as f:
        for line in f:
            number_branches += 1
            pc, outcome = line.split()
            pc = int(pc, 16)  # Convert PC from hexadecimal to integer

            prediction = predictor.predict(pc)  # Get prediction
            predictor.update(outcome)  # Update predictor
            if prediction != outcome:
                number_mispredictions += 1  # Count mispredictions

    # Print results
    print("number of predictions: ", number_branches)
    print("number of mispredictions: ", number_mispredictions)
    print(f"misprediction rate: {(number_mispredictions / number_branches) * 100:.2f}%")

    # Print final contents of the predictor
    if args.predictor == "smith":
        print(f"FINAL COUNTER CONTENT: {predictor.counter}")
    elif args.predictor == "bimodal":
        print("FINAL BIMODAL CONTENTS")
        for i, entry in enumerate(predictor.table):
            print(f"{i} {entry}")
    elif args.predictor == "gshare":
        print("FINAL GSHARE CONTENTS")
        for i, entry in enumerate(predictor.table):
            print(f"{i} {entry}")
    elif args.predictor == "hybrid":
        print("FINAL CHOOSER CONTENTS")
        for i, entry in enumerate(predictor.chooser_table):
            print(f"{i} {entry}")
        print("FINAL GSHARE CONTENTS")
        for i, entry in enumerate(predictor.gshare.table):
            print(f"{i} {entry}")
        print("FINAL BIMODAL CONTENTS")
        for i, entry in enumerate(predictor.bimodal.table):
            print(f"{i} {entry}")

File: sim.cc
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <cstdlib>
#include <iomanip>

// Smith predictor: A simple counter-based predictor.
// It uses a saturating counter to predict whether a branch will be taken ("t") or not taken ("n").
class Smith {
private:
    int b; // Number of bits in the counter
public:
    int counter; // Current value of the counter
    int thres;   // Threshold for predicting "t"

    Smith(int bVal) : b(bVal) {
        // Initialize counter to mid-point and set the threshold
        counter = 1 << (b - 1);
        thres = 1 << (b - 1);
    }

    std::string predict() {
        // Predict "t" if counter >= threshold, otherwise predict "n"
        return (counter >= thres) ? "t" : "n";
    }

    void update(const std::string &outcome) {
        // Update the counter based on the actual outcome
        if (outcome == "t") {
            counter = std::min(counter + 1, (1 << b) - 1); // Increment counter
        } else {
            counter = std::max(counter - 1, 0); // Decrement counter
        }
    }
};

// Bimodal predictor: Uses a table of counters indexed by the program counter (PC).
// Each counter predicts "t" or "n" based on its value.
class Bimodal {
private:
    int m; // Number of index bits
public:
    std::vector<int> table; // Table of counters
    int index;              // Current index in the table

    Bimodal(int mVal) : m(mVal), index(0) {
        // Initialize the table with mid-point values
        table.resize(1 << m, 4);
    }

    std::string predict(unsigned pc) {
        // Calculate the index using bits m+1 to 2 of the PC
        index = (pc >> 2) & ((1 << m) - 1);
        // Predict "t" if counter >= 4, otherwise predict "n"
        return (table[index] >= 4) ? "t" : "n";
    }

    void update(const std::string &outcome) {
        // Update the counter at the calculated index
        if (outcome == "t") {
            table[index] = std::min(table[index] + 1, 7); // Increment counter
        } else {
            table[index] = std::max(table[index] - 1, 0); // Decrement counter
        }
    }
};

// Gshare predictor: Combines global history with the PC to index into a table of counters.
class Gshare {
private:
    int m, n; // m: Number of index bits, n: Number of global history bits
public:
    int getN() const { return n; }
public:
    std::vector<int> table;    // Table of counters
    unsigned global_history;   // Global history register
    int idx;                   // Current index in the table

    Gshare(int mVal, int nVal) : m(mVal), n(nVal), global_history(0), idx(0) {
        // Initialize the table with mid-point values
        table.resize(1 << m, 4);
    }

    std::string predict(unsigned pc) {
        // Calculate the index by XORing the PC and global history
        idx = ((pc >> 2) & ((1 << m) - 1)) ^ global_history;
        // Predict "t" if counter >= 4, otherwise predict "n"
        return (table[idx] >= 4) ? "t" : "n";
    }

    void update(const std::string &outcome) {
        // Update the counter at the calculated index
        if (outcome == "t") {
            table[idx] = std::min(table[idx] + 1, 7); // Increment counter
        } else {
            table[idx] = std::max(table[idx] - 1, 0); // Decrement counter
        }
        // Update the global history
        global_history = (global_history >> 1) | ((outcome == "t") << (n - 1));
    }
};

// Hybrid predictor: Combines Gshare and Bimodal predictors using a chooser table.
class Hybrid {
private:
    int k; // Number of chooser table index bits
public:
    Gshare gshare;                  // Gshare predictor
    Bimodal bimodal;                // Bimodal predictor
    std::vector<int> chooser_table; // Chooser table
    int idx;                        // Current index in the chooser table
    std::string gshare_output;      // Gshare prediction
    std::string bimodal_output;     // Bimodal prediction

    Hybrid(int kVal, int m1, int n, int m2)
     : k(kVal), gshare(m1, n), bimodal(m2), idx(0) {
        // Initialize chooser table with mid-point values
        chooser_table.resize(1 << k, 1);
    }

    std::string predict(unsigned pc) {
        // Calculate the chooser table index
        idx = (pc >> 2) & ((1 << k) - 1);
        // Get predictions from Gshare and Bimodal
        gshare_output = gshare.predict(pc);
        bimodal_output = bimodal.predict(pc);
        // Use chooser table to select the predictor
        return (chooser_table[idx] >= 2) ? gshare_output : bimodal_output;
    }

    void update(const std::string &outcome) {
        // Update the predictor used
        if (chooser_table[idx] >= 2) {
            gshare.update(outcome);
        } else {
            bimodal.update(outcome);
            // Also update the global history of the Gshare predictor
            gshare.global_history =
               (gshare.global_history >> 1) |
               ((outcome == "t") << (gshare.getN() - 1));
        }
        // Update the chooser table based on prediction accuracy
        if (gshare_output == outcome && bimodal_output != outcome) {
            chooser_table[idx] = std::min(chooser_table[idx] + 1, 3);
        } else if (gshare_output != outcome && bimodal_output == outcome) {
            chooser_table[idx] = std::max(chooser_table[idx] - 1, 0);
        }
    }
};

int main(int argc, char* argv[]) {
    // Parse command line arguments
    if (argc < 3) {
        std::cerr << "Invalid arguments.\n";
        return 1;
    }

    std::string predictorType = argv[1];
    std::string arg1 = argv[2], arg2, arg3, arg4, arg5;
    if (argc > 3) arg2 = argv[3];
    if (argc > 4) arg3 = argv[4];
    if (argc > 5) arg4 = argv[5];
    if (argc > 6) arg5 = argv[6];

    // Initialize predictor objects based on the type
    Smith* smithPtr = nullptr;
    Bimodal* bimodalPtr = nullptr;
    Gshare* gsharePtr = nullptr;
    Hybrid* hybridPtr = nullptr;
    std::string trace_file;

    if (predictorType == "smith") {
        smithPtr = new Smith(std::stoi(arg1));
        trace_file = arg2;
    } else if (predictorType == "gshare") {
        gsharePtr = new Gshare(std::stoi(arg1), std::stoi(arg2));
        trace_file = arg3;
    } else if (predictorType == "bimodal") {
        bimodalPtr = new Bimodal(std::stoi(arg1));
        trace_file = arg2;
    } else if (predictorType == "hybrid") {
        hybridPtr = new Hybrid(std::stoi(arg1),
                               std::stoi(arg2),
                               std::stoi(arg3),
                               std::stoi(arg4));
        trace_file = arg5;
    } else {
        std::cout << "Invalid predictor type\n";
        return 1;
    }

    // Open the trace file
    std::ifstream fin(trace_file);
    if (!fin.is_open()) {
        std::cerr << "Could not open trace file.\n";
        return 1;
    }

    // Print command line information
    std::cout << "COMMAND\n";
    std::cout << "./sim";
    for (int i = 1; i < argc; i++) {
        std::cout << " " << argv[i];
    }
    std::cout << "\n";
    std::cout << "OUTPUT\n";

    // Process the trace file
    int number_branches = 0;
    int number_mispredictions = 0;
    std::string line;
    while (std::getline(fin, line)) {
        if (line.empty()) continue;
        number_branches++;
        std::stringstream ss(line);
        std::string pcHex, outcome;
        ss >> pcHex >> outcome;
        unsigned pcVal = std::stoul(pcHex, nullptr, 16);

        std::string prediction;
        if (smithPtr) {
            prediction = smithPtr->predict();
            smithPtr->update(outcome);
        } else if (bimodalPtr) {
            prediction = bimodalPtr->predict(pcVal);
            bimodalPtr->update(outcome);
        } else if (gsharePtr) {
            prediction = gsharePtr->predict(pcVal);
            gsharePtr->update(outcome);
        } else if (hybridPtr) {
            prediction = hybridPtr->predict(pcVal);
            hybridPtr->update(outcome);
        }
        if (prediction != outcome) {
            number_mispredictions++;
        }
    }
    fin.close();

    // Print statistics
    std::cout << "number of predictions: " << number_branches << "\n";
    std::cout << "number of mispredictions: " << number_mispredictions << "\n";
    double rate = (number_branches == 0 ? 0.0
                  : 100.0 * number_mispredictions / number_branches);
    std::cout << "misprediction rate: " << std::fixed << std::setprecision(2)
              << rate << "%\n";

    // Print final contents of the predictor
    if (smithPtr) {
        std::cout << "FINAL COUNTER CONTENT: " << smithPtr->counter << "\n";
    } else if (bimodalPtr) {
        std::cout << "FINAL BIMODAL CONTENTS\n";
        for (size_t i = 0; i < bimodalPtr->table.size(); i++) {
            std::cout << i << " " << bimodalPtr->table[i] << "\n";
        }
    } else if (gsharePtr) {
        std::cout << "FINAL GSHARE CONTENTS\n";
        for (size_t i = 0; i < gsharePtr->table.size(); i++) {
            std::cout << i << " " << gsharePtr->table[i] << "\n";
        }
    } else if (hybridPtr) {
        std::cout << "FINAL CHOOSER CONTENTS\n";
        for (size_t i = 0; i < hybridPtr->chooser_table.size(); i++) {
            std::cout << i << " " << hybridPtr->chooser_table[i] << "\n";
        }
        std::cout << "FINAL GSHARE CONTENTS\n";
        for (size_t i = 0; i < hybridPtr->gshare.table.size(); i++) {
            std::cout << i << " " << hybridPtr->gshare.table[i] << "\n";
        }
        std::cout << "FINAL BIMODAL CONTENTS\n";
        for (size_t i = 0; i < hybridPtr->bimodal.table.size(); i++) {
            std::cout << i << " " << hybridPtr->bimodal.table[i] << "\n";
        }
    }

    // Clean up dynamically allocated memory
    delete smithPtr;
    delete bimodalPtr;
    delete gsharePtr;
    delete hybridPtr;
    return 0;
}